### PR TITLE
[ISSUE #682] Fix wrong message trace detail 

### PR DESCRIPTION
**Pull request by maixiaohai**
## What is the purpose of the change

issue #682 

## Brief changelog

1. add client host field
2. add compatbile logic when decode trace msg

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: rocketmq-console/src/main/java/org/apache/rocketmq/console/util/MsgTraceDecodeUtil.java [hunk 1]
```diff
@@ -92,6 +100,11 @@
                     // add the context type
                     subAfterContext.setContextCode(Integer.parseInt(line[6]));
                 }
+                // compatible with the old version
+                if (line.length >= 9) {
```
Comment (vongosling) [inline rocketmq-console/src/main/java/org/apache/rocketmq/console/util/MsgTraceDecodeUtil.java:104]: Could we use the constant while not a number(such as 9 or 15)to be forward compatible?

**Review by vongosling — CHANGES_REQUESTED** [review]

File: rocketmq-console/src/main/java/org/apache/rocketmq/console/util/MsgTraceDecodeUtil.java [hunk 2]
```diff
@@ -54,12 +66,29 @@
...
 
-                if (line.length == 13) {
-                    pubContext.setSuccess(Boolean.parseBoolean(line[12]));
-                } else if (line.length == 14) {
-                    bean.setOffsetMsgId(line[12]);
-                    pubContext.setSuccess(Boolean.parseBoolean(line[13]));
+                // compatible with different version
+                switch (line.length) {
+                    case TRACE_MSG_PUB_V1_LEN:
+                        break;
```
Comment (vongosling) [inline rocketmq-console/src/main/java/org/apache/rocketmq/console/util/MsgTraceDecodeUtil.java:72]: Make some tests for the switch case?

**Review by vongosling — CHANGES_REQUESTED** [review]

**Comment by maixiaohai** [comment]
soft ping~

File: rocketmq-console/src/main/java/org/apache/rocketmq/console/util/MsgTraceDecodeUtil.java [hunk 3]
```diff
@@ -88,9 +118,24 @@
...
+                    case TRACE_MSG_SUBAFTER_V2_LEN:
+                        subAfterContext.setContextCode(Integer.parseInt(line[6]));
+                        break;
+                    case TRACE_MSG_SUBAFTER_V3_LEN:
+                        subAfterContext.setContextCode(Integer.parseInt(line[6]));
+                        subAfterContext.setTimeStamp(Long.parseLong(line[7]));
+                        subAfterContext.setGroupName(line[8]);
+                        break;
+                    default:
```
Comment (vongosling) [inline rocketmq-console/src/main/java/org/apache/rocketmq/console/util/MsgTraceDecodeUtil.java:133]: What's the difference between TRACE_MSG_SUBAFTER_V1_LEN and default logic, It is correct here?

**Review by vongosling — CHANGES_REQUESTED** [review]

File: rocketmq-console/src/main/java/org/apache/rocketmq/console/util/MsgTraceDecodeUtil.java (hunk 3 above)
Comment (maixiaohai) [reply rocketmq-console/src/main/java/org/apache/rocketmq/console/util/MsgTraceDecodeUtil.java:133]: default logic is compatible for new version subAfterType.
If a new field is added to subAfter message, the length will greater than TRACE_MSG_SUBAFTER_V3_LEN.

**Comment by vongosling** [comment]
Great, TraceDataEncoder still need to polish ~

**Pull request merged**